Add conformance test for deferred shader compilations with extensions.

DavidYen opened the pull request:
Add conformance test for deferred shader compilations with extensions.

If shader compilation is deferred until the link time, a test has been added to be sure the shader compiler uses its state as seen during the compilation request, not the link request. The added conformance test does the following to test this:

1. Create a new GL context which does not have the tested extension enabled.
2. Compile the vertex and fragment shaders which would fail if the extension were disabled.
3. Enable the extension by querying if the extension is supported.
4. Link the shader and check the link status, this should fail since at the time of the compilation the extension was disabled.

zhenyao commented:
LGTM

zhenyao commented on sdk/tests/conformance/extensions/ext-frag-depth.html:
```diff
@@ -116,6 +116,9 @@
 runAllTests();
 
+// Run deferred link tests.
+runDeferredLinkTests();
```
nit: this should be part of the runAllTests(), here and below.

kenrussell replied:
This is an excellent point -- the new runDeferredLinkTests function assumes the extension is supported. It's perfectly legal for it to not be. Adding this to the list of tests in runAllTests() in this and the other extension tests will implicitly fix this problem.

toji commented on sdk/tests/conformance/extensions/ext-frag-depth.html:
```diff
@@ -244,6 +247,66 @@
+
+    var linked = gl_deferred.getProgramParameter(program, gl_deferred.LINK_STATUS);
+    if (linked) {
+        testFailed("Compilation with extension disabled then linking with extension enabled caused successful compile.");
```
This is a slightly awkward failure message. Might be more natural to say "Compilation with extension disabled then linking with extension enabled should have failed."?

kenrussell commented:
Thanks for this contribution. Once Brandon's comments are addressed this looks good to me.

One suggestion: it might be worth adding another optional argument to loadShader in sdk/tests/conformance/resources/webgl-test-utils.js like "opt_ignoreCompileStatus" which, if set to true, just skips the query of the shader's COMPILE_STATUS. A lot of boilerplate code could then be removed from these new tests.

kenrussell commented:
Sorry, I should have said "Brandon's and Mo's comments".

There's one more issue which I'll detail in the diffs.

DavidYen commented:
I have refactored the load shader functions in the utility file to have an option to skip the compilation status check. With this change I also added a check to be sure the shader compiles as expected in a separate context before we do the deferred test.

The reason I had originally placed the new deferred link test outside of RunAllTests() is because it uses it's own context and not the GL context that the RunAllTests() function utilizes. Since my test is checking if the shader links, the result is the same whether the extension is supported or not. I think it is a bit confusing to have 1 test outside of "RunAllTests()" though so I just placed it in there to make it match all the other tests. The only downside is that it will run twice unnecessarily, but that is not a big deal for this test.

zhenyao commented:
LGTM

kenrussell commented on sdk/tests/conformance/extensions/ext-frag-depth.html:
```diff
@@ -140,6 +140,9 @@
     }
+
+    // Run deferred link tests.
+    runDeferredLinkTests();
```
This should have been added in the block above where EXT_frag_depth was successfully enabled. This test can't run properly if the extension isn't supported. See below.

kenrussell commented on sdk/tests/conformance/extensions/ext-frag-depth.html:
```diff
@@ -244,6 +247,62 @@
+
+    // First test if link succeeds with extension enabled.
+    var glEnabled = wtu.create3DContext();
+    var extEnabled = glEnabled.getExtension("EXT_frag_depth");
```
EXT_frag_depth might not be supported by the WebGL implementation. In this case, the call to getExtension will fail.

kenrussell commented on sdk/tests/conformance/extensions/ext-frag-depth.html:
```diff
@@ -244,6 +247,62 @@
+    var extEnabled = glEnabled.getExtension("EXT_frag_depth");
+
+    var vertexShader = wtu.loadShaderFromScript(glEnabled, "goodVertexShader");
+    var fragmentShader = wtu.loadShaderFromScript(glEnabled, "macroFragmentShader");
```
If the above call to getExtension failed, the macroFragmentShader won't compile successfully; GL_EXT_frag_depth will not be defined, and the #error directive will be hit.

kenrussell commented on sdk/tests/conformance/extensions/ext-frag-depth.html:
```diff
@@ -244,6 +247,62 @@
+    var vertexShader = wtu.loadShaderFromScript(glEnabled, "goodVertexShader");
+    var fragmentShader = wtu.loadShaderFromScript(glEnabled, "macroFragmentShader");
+
+    if (!vertexShader || !fragmentShader) {
```
fragmentShader will be null here if compilation happens eagerly in the WebGL implementation and EXT_frag_depth wasn't supported.
